code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,513 @@
+			to_chat(src, span_boldwarning("You cannot send IC messages (muted)."))
+			return
+		if (!(ignore_spam || forced) && src.client.handle_spam_prevention(message, MUTE_IC))
+			return
+
+	if(sanitize)
+		message = trim(copytext_char(sanitize(message), 1, MAX_MESSAGE_LEN))
+
+	src.log_talk(message, LOG_SAY)
+	var/rendered = span_deadsay("<b>UNDEAD: [src]</b> says, \"[message]\"")
+	relay_to_list_and_observers(rendered, GLOB.revenant_relay_mobs, src)
+
+/mob/living/basic/revenant/ClickOn(atom/A, params) //revenants can't interact with the world directly, so we gotta do some wacky override stuff
+	var/list/modifiers = params2list(params)
+	if(LAZYACCESS(modifiers, SHIFT_CLICK))
+		ShiftClickOn(A)
+		return
+	if(LAZYACCESS(modifiers, ALT_CLICK))
+		AltClickNoInteract(src, A)
+		return
+	if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		ranged_secondary_attack(A, modifiers)
+		return
+
+	if(ishuman(A) && in_range(src, A))
+		attempt_harvest(A)
+
+/mob/living/basic/revenant/ranged_secondary_attack(atom/target, modifiers)
+	if(revealed || inhibited || HAS_TRAIT(src, TRAIT_NO_TRANSFORM) || !Adjacent(target) || !incorporeal_move_check(target))
+		return
+
+	var/list/icon_dimensions = get_icon_dimensions(target.icon)
+	var/orbitsize = (icon_dimensions["width"] + icon_dimensions["height"]) * 0.5
+	orbitsize -= (orbitsize / world.icon_size) * (world.icon_size * 0.25)
+	orbit(target, orbitsize)
+
+/mob/living/basic/revenant/death()
+	if(!revealed || dormant) //Revenants cannot die if they aren't revealed //or are already dead
+		return
+	ADD_TRAIT(src, TRAIT_NO_TRANSFORM, REVENANT_STUNNED_TRAIT)
+	dormant = TRUE
+
+	visible_message(
+		span_warning("[src] lets out a waning screech as violet mist swirls around its dissolving body!"),
+		span_revendanger("NO! No... it's too late, you can feel your essence [pick("breaking apart", "drifting away")]..."),
+	)
+
+	revealed = TRUE
+	invisibility = 0
+	icon_state = "revenant_draining"
+	playsound(src, 'sound/effects/screech.ogg', 100, TRUE)
+
+	for(var/i = alpha, i > 0, i -= 10)
+		alpha = i
+		CHECK_TICK
+
+	visible_message(span_danger("[src]'s body breaks apart into a fine pile of blue dust."))
+
+	var/obj/item/ectoplasm/revenant/goop = new(get_turf(src))
+	goop.old_ckey = client.ckey //If the essence reforms, the old revenant is put back in the body
+	goop.revenant = src
+	revealed = FALSE
+	forceMove(goop)
+
+/mob/living/basic/revenant/adjust_health(amount, updating_health = TRUE, forced = FALSE)
+	if(!forced && !revealed)
+		return 0
+
+	. = amount
+
+	essence = max(0, essence - amount)
+	if(updating_health)
+		update_health_hud()
+	if(essence == 0)
+		death()
+
+	return .
+
+/mob/living/basic/revenant/orbit(atom/target)
+	setDir(SOUTH) // reset dir so the right directional sprites show up
+	return ..()
+
+/mob/living/basic/revenant/stop_orbit(datum/component/orbiter/orbits)
+	// reset the simple_flying animation
+	animate(src, pixel_y = 2, time = 1 SECONDS, loop = -1, flags = ANIMATION_RELATIVE)
```

Comment:
Do we still need this? Looks like legacy code which might now be handled by the `SIGNAL_ADDTRAIT(TRAIT_NO_FLOATING_ANIM) / SIGNAL_REMOVETRAIT(TRAIT_NO_FLOATING_ANIM)` in `/datum/element/movetype_handler`

Reply:
i'll have to look at that